Finalizing `output()` and `outputFromObservable()` APIs

devversion opened the pull request:
See individual commits.

———

### `output()`

This commit exposes the new `output()` API with numerous benefits:

- Symmetrical API to `input()`, `model()` etc.
- Fixed types for `EventEmitter.emit`— current `emit` method of
  `EventEmitter` is broken and accepts `undefined` via `emit(value?: T)`
- Removal of RxJS specific concepts from outputs. error channels,
  completion channels etc. We now have a simple consistent
  interface.
- Automatic clean-up of subscribers upon directive/component destory-
  when subscribed programmatically.

```ts
import {output} from '@angular/core';

@Directive({..})
export class MyDir {
  nameChange = output<string>();     // OutputEmitterRef<string>
  onClick = output();                // OutputEmitterRef<void>

  doSomething() {
    this.onClick.emit();
  }
}
```

Note: RxJS custom observable cases will be handled in future commits via
explicit helpers from the interop.

### `outputFromObservable()`

```ts
import {outputFromObservable} from '@angular/core/rxjs-interop';

 @Directive({..})
 export class MyDir {
   nameChange$ = new BehavorSubject<string>();
   nameChange = outputFromObservable(this.nameChange$); // OutputRef<string>
}
```

### `outputToObservable`

```ts
import {outputToObservable} from '@angular/core/rxjs-interop';


outputToObservable(myInstance.someOutput)
  .pipe(take(1))
  .subscribe(...)
```

devversion committed:
fix(compiler-cli): support `ModuleWithProviders` literal detection with `typeof`

As part of improving test safety of the compiler, I've noticed that
we have a special pass for detecting external `ModuleWithProviders`
where we detect the module type from an object literal.

This literal is structured like the following: `{ngModule: T}`. The
detection currently takes `T` directly, but in practice it should be
`typeof T` to satisfy the `ModuleWithProviders` type that is accepted
as part of `Component#imports`.

This commit adds support for this, so that we can fix the unit test
in preparation for using the real Angular core types in ngtsc tests.

devversion committed:
test: replace `fake_core` with real `@angular/core` output

This commit replaces `fake_core` with the real `@angular/core`
output. See previous commit for reasons.

Overall, this commit:

* Replaces references of `fake_core`
* Fixes tests that were testing Angular compiler detection that _would_
  already be flagged by type-checking of TS directly. We keep these
  tests for now, and add `@ts-ignore` to verify the Angular checks, in
  case type checking is disabled in user applications- but it's worth
  considering to remove these tests. Follow-up question/non-priority.
* Adds `@ts-ignore` to the tests for `defer` 1P because the property is
  marked as `@internal` and now is (correctly) causing failures in the
  compiler test environment.
* Fixes a couple of tests with typos, wrong properties etc that
  previously weren't detected! A good sign.

devversion committed:
test: integration test for `output()` api

Adds an integration test for `output()` and `outputFromObservable()`.
The test verifies the JIT transform as well.

devversion committed:
refactor(compiler-cli): support detecting initializer APIs from different modules

This commit allows us to detect initializer APIs like
`outputFromObservable` that are declared in different modules- not
necessarily `@angular/core`.

devversion committed:
refactor(core): create `OutputRef` runtime construct

This commit creates the proposed `OutputRef` interface along
with `OutputEmitterRef`:

- `OutputRef` is the consistent interface for all Angular outputs.
- `OutputEmitterRef` is an extension for emitting values. Like
  `EventEmitter`.
- subscriptions are cleaned up automatically upon directive/component
  destroy.
- emitting is an error when destroyed
- subscribing programmatically is an error when already destroyed.

This commit will also implement the shared output ref runtime construct,
that can be used by `output()`, `outputFromObservable()` and `model()`.

We will manage subscriptions in a simple way, manually, without RxJS.

devversion committed:
test: add compliance generated output tests for `output()`

This commit adds compliance tests for the new output APIs.

devversion committed:
test: add language-service tests for `outputFromObservable()`

Adds additional language-service tests for `outputFromObservable()`.
Existing tests already verify the behavior for `output()`.

devversion committed:
test: fix global augmentation leaking into all source files

The `inject` global augmentation from upgrade tests, leak into
all source files for IDEs, making it easy to run into issues
when actually trying to deal with `inject` from Angular core for DI.

devversion committed:
test: add compiler ngtsc tests for new `outputFromObservable` API

Adds additional ngtsc compiler tests for the `outputFromObservable` API.

devversion committed:
feat(core): introduce `outputToObservable` interop helper

This commit introduces an addition to `output()` and
`outputFromObservable`()` —called `outputToObservable()`.

The helper lives in the RxJS interop package and allows agnostic
programmatic subscriptions to `OutputRef`s by converting the output
to an observable with `.pipe` etc.

The function is ideally used in all places where you subscribe to an
output programmatically. Those outputs in the future, with the new APIs,
may not be actual RxJS constructs, but abstract `OutputRef`'s that
simply expose a `.subscribe` method. The helper allows you to
agnostically convert outputs to RxJS observables that you can safely
interact with.

The observables are also completed automatically, if possible, when the
owning directive/component is destroyed— Something that is not
guaranteed right now.

devversion committed:
refactor(core): `model()` implements `OutputRef`

A model signal is technically an output, at runtime and conceptually.

This commit re-uses the shared output ref logic and ensures the
interfaces match.

devversion committed:
refactor(core): `EventEmitter` implements `OutputRef`.

An `EventEmitter` is a construct owned by Angular that should be
used for outputs as of right now.

As we are introducing the new `OutputRef` interface for the new output
function APIs, we also think `EventEmitter` should implement
`OutputRef`— ensuring all "known" outputs follow the same contract.

This commit ensures `EventEmitter` implements an `OutputRef`

Note: An output ref captures the destroy ref from the current injection
context for clean-up purposes. This is also done for `EventEmitter` in a
backwards compatible way:

- not requiring an injection context. EventEmitter may be used
  elsewhere.
- not cleaning up subscriptions/completing the emitter when the
  directive/component is destroyed. This would be a change in behavior.

Note 2: The dependency on `DestroyRef` causes it to be retained in all
bundling examples because ironically `NgZone` uses `EventEmitter`- not
for outputs. The code is pretty minimal though, so that should be
acceptable.

`EventEmitter` will now always retain `NgZone. This increases the
payload size slightly around 800b for AIO. Note that the other increases
were coming from previous changes. This commit just pushed it over the
threshold.

devversion committed:
test: add runtime tests for `output()` function API

Adds runtime acceptance tests for `output()` and
`outputFromObservable()`.

devversion committed:
refactor(core): enforce `model()` and `output()` is used in an injection context

Technically `model()` and `output()` already need to be defined in an
injection context- because `OutputRef` requires this.

To improve the error messaging, this commit asserts this as part of the
top-level entry functions for `model()` and `output()`. Without this
change, the error would mention the `_createOutputRef` internal
function.

devversion committed:
feat(core): introduce `outputFromObservable()` interop function

Introduces a second API in addition to the new `output()` function.

The new function `outputFromObservable()` can be used to declare outputs
using the new `OutputRef` API and `output()` API, while using a custom
RxJS observable as data source.

This is something that is currently possible in Angular and we would
like to keep possible- even though we never intended to support custom
observables aside from RxJS-based `EventEmitter`.

The interop bridges the gap and allows you to continue using
`Subject`, `ReplaySubject`, `BehaivorSubjct,` - or cold custom
observables for outputs. You can still trigger logic only when
the output is subscribed- unlike with imperative `emit`s of
`EventEmitter` or the new `OutputEmitterRef`.

A notable difference is that you need two class members where you
previously could access the `Subject` directly. This is an intentional
trade-off we've made to ensure that all new outputs implement the
`OutputRef` interface and we are exposing a minimal API surface to
consumers of components that currently access the output
programmatically.

devversion committed:
feat(core): expose new `output()` API

This commit exposes the new `output()` API with numerous benefits:

- Symmetrical API to `input()`, `model()` etc.
- Fixed types for `EventEmitter.emit`— current `emit` method of
  `EventEmitter` is broken and accepts `undefined` via `emit(value?: T)`
- Removal of RxJS specific concepts from outputs. error channels,
  completion channels etc. We now have a simple consistent
  interface.
- Automatic clean-up of subscribers upon directive/component destory-
  when subscribed programmatically.

```ts
@Directive({..})
export class MyDir {
  nameChange = output<string>();     // OutputEmitterRef<string>
  onClick = output();                // OutputEmitterRef<void>
}
```

Note: RxJS custom observable cases will be handled in future commits via
explicit helpers from the interop.

devversion committed:
refactor(compiler-cli): speed up compiler tests by caching Angular

Currently the `makeProgram` utility from `ngtsc/testing` does not use
the test host by default- optimizing for source file caching.

Additionally, the host can be updated to attempt caching of the `.d.ts`
files from `@angular/core`— whether that's fake core, or the real core-
is irrelevant. We are never caching if these changes between tests, so
correctness is guaranteed.

This commit reduces the type check test times form 80s to just 11
seconds, faster than what it was before with `fake_core`. The ngtsc
tests also run significantly faster. From 40s to 30s

markostanimirovic commented:
A great addition to the `rxjs-interop` plugin! 🚀  However, function names seem too verbose to me. Consider the following naming:

- ~`outputFromObservable`~ => `rxOutput`
- ~`outputToObservable`~ => add an additional signature to the existing `toObservable` function that accepts `OutputRef` as input argument

```diff
-import { outputFromObservable } from '@angular/core/rxjs-interop';
+import { rxOutput } from '@angular/core/rxjs-interop';

@Directive({..})
export class MyDir {
  nameChange$ = new Subject<string>();
- nameChange = outputFromObservable(this.nameChange$);
+ nameChange = rxOutput(this.nameChange$);
}
```

alxhub commented on packages/core/rxjs-interop/src/output_to_observable.ts:
```diff
@@ -0,0 +1,35 @@
+  });
+
+  // Avoid listening multiple times as we can share the events.
+  return observable.pipe(share());
```
I don't see a reason to `share` here - there's no harm in having multiple subscriptions to the `OutputRef`. `share` is an expensive operator (code-size wise) and we try to avoid introducing new operator dependencies when possible.

devversion replied:
Yeah, no strong reason, except I didn't see a reason we wouldn't want to share subscriptions here. Removing as code-side is a fair argument!

alxhub commented on packages/core/src/event_emitter.ts:
```diff
@@ -8,6 +8,10 @@
 import {PartialObserver, Subject, Subscription} from 'rxjs';
 
+import {OutputRef, OutputRefUnsubscribeFn} from './authoring/output/output_ref';
+import {inject} from './di';
```
Nit: prefer deep imports rather than these top-level barrel files. They're well known for causing circular deps.

Ideally they wouldn't, but our package structure isn't great.

devversion replied:
Changing. I faced a couple ones, but resolved them as per the checks— which wasn't too diffuclt.

alxhub commented on packages/core/src/event_emitter.ts:
```diff
@@ -97,22 +101,32 @@ export interface EventEmitter<T> extends Subject<T> {
+    // Attempt to retrieve a `DestroyRef` optionally.
+    // For backwards compatibility reasons, this cannot be required.
+    try {
+      this.__internal_DestroyRef = inject(DestroyRef);
```
Why not `inject(DestroyRef, {optional: true})`?

Or are you worried about `EventEmitter`s created outside of DI contexts? I would recommend avoiding the `try`/`catch` and checking `getInjectImplementation() || getCurrentInjector()` (see impl of `assertInInjectionContext`.

Also, have we checked in g3 whether making `EventEmitter` depend on injection context is that breaking?

devversion replied:
It's because `EventEmitter` may be constructed outside of an injection context, yes.

Started a TGP to check how breaking it would be.

devversion replied:
Results: 40k failures, unfeasible. Most issues arise from `QueryList` lazily creating an `EventEmitter` (?!) lazily for `.changes`. This can happen outside of injection contexts and affects a lot of targets

alxhub commented on packages/core/src/event_emitter.ts:
```diff
@@ -142,7 +156,9 @@ class EventEmitter_ extends Subject<any> {
     }
 
-    return sink;
+    return new Proxy(sink, {
```
😲 

Clever.

crisbeto commented on packages/core/rxjs-interop/src/output_from_observable.ts:
```diff
@@ -0,0 +1,61 @@
+
+      // Stop yielding more values when the directive/component is already destroyed.
+      const subscription =
+          observable.pipe(takeUntil(onDestroy)).subscribe(value => callbackFn(value));
```
Do we need to do something with the error/complete channels here?

devversion replied:
Good point. I think we should just document that users are responsible for handling errors beforehand. If they aren't, the error will be simply handled by RxJS and will be reported unhandled to `window.onerror` (see RxJS` reportUnhandledError.ts`)

crisbeto commented on packages/core/src/authoring/model/model.ts:
```diff
@@ -6,15 +6,20 @@
 import {createModelSignal, ModelOptions, ModelSignal} from './model_signal';
 
 export function modelFunction<T>(initialValue?: T): ModelSignal<T|undefined> {
+  assertInInjectionContext(model);
```
Should these be guarded with `ngDevMode`? Also they shouldn't be necessary once we implement the compiler diagnostic that prevents the function from being used outside of initializers, but it doesn't hurt it either.

crisbeto commented on packages/core/src/authoring/model/model_signal.ts:
```diff
@@ -85,23 +77,17 @@ export function createModelSignal<T>(initialValue: T): ModelSignal<T> {
   }
 
-  return getter as ModelSignal<T>;
+  return getter as (typeof getter&Pick<
```
At this point I feel like `as any` might be an improvement 😄

devversion replied:
Haha, this is 100% type safe though and ensures we are actually setting what is needed (compared to just casting)

crisbeto commented on packages/core/src/authoring/output/output_ref.ts:
```diff
@@ -0,0 +1,157 @@
+   * Angular will automatically clean up the subscription when
+   * the directive/component of the output is destroyed.
+   */
+  subscribe(cb: (value: T) => void): OutputRefUnsubscribeFn;
```
Given that it's a public API, maybe we should avoid the abbreviations in the names?

devversion replied:
I did have `listenerFunction`, but it ended up making the IDE hints rather unreadable. Happy to rename

crisbeto commented on packages/core/src/authoring/output/output_ref.ts:
```diff
@@ -0,0 +1,157 @@
+}) {
+  // Outputs refs are expected to be created in injection contexts.
+  // I.e. during component instance creation.
+  assertInInjectionContext(createOutputRef);
```
Guard this with `ngDevMode`?

crisbeto commented on packages/core/src/authoring/output/output_ref.ts:
```diff
@@ -0,0 +1,157 @@
+}
+
+/** Gets the owning `DestroyRef` for the given output. */
+export function getOutputDestroyRef(ref: OutputRef<unknown>): DestroyRef|undefined {
```
Should we just read the property directly?

devversion replied:
We can't because in the `interop` the `.destroyRef` is `@internal` and gone (due to separate TS compilations). This gives us some type safety here

devversion commented on packages/compiler-cli/src/ngtsc/annotations/directive/src/initializer_functions.ts:
```diff
@@ -24,16 +23,19 @@ import {CORE_MODULE} from '../../common';
-export type InitializerApiFunction =
-    'input'|'model'|'ɵoutput'|'output'|'viewChild'|'viewChildren'|'contentChild'|'contentChildren';
+export interface InitializerApiFunction {
+  owningModule: '@angular/core'|'@angular/core/rxjs-interop';
```
Yeah, I think this is fine, but happy to change if you feel strongly. The idea is to just give some insight on what APIs are supported initializer functions

pkozlowski-opensource reviewed (approved):
LGTM. As discussed in person I'm not sure about the completion semantics of the output as Observable and - more importantly - `EventEmitter` involvement in it. I would lean towards dropping the completion logic but at the same time I don't feel like I understand all the consequences here.

Reviewed-for: public-api 
Reviewed-for: fw-core

brandonroberts commented:
I agree with `toOutput` to be consistent with the other helper functions. This is the other direction, but I recall Ben Lesh providing some feedback against using the `fromThing` convention for naming.

Also from an import perspective, you might hit `output` much more frequently when trying to get to `outputFromObservable` when using autocomplete because of the naming similarity. `toOutput` being similar to the others is more intentional when trying to use it.

devversion committed:
refactor(core): revert listener subscription typeof check

For model signals we introduced some sniffing on the return type of a
`.subscribe` invocation- allowing for subscribe to _just_ return a
callback directly to unsubscribe.

This works in practice, but the positive `tCleanup` indices have more
meaning, especially in the context of `DebugElement`. A positive index
indicates a DOM event- so we need to revert this change. This now
surfaced as we made `EventEmitter` return a function + the subscription
via a proxy that ended up `typeof function` --> and broke some tests
where debug element incorrectly invoked non-dom outputs as dom
listeners. We don't need this change with current unsubscribe function
concept.

devversion commented:
Capturing this context. I've reached out to GDEs to discuss the naming further. We had a couple of interesting points, but overall the consensus is that `observableFromOutput` is what we want, mostly because:

- 97% of outputs in the google codebase will not need to use the interop— so it's unclear it warrants a super short name. You can alias if you want (locally in the file)
- `toOutput` is different to `outputFromObservable` as it indicates more of a runtime conversion helper. Like `toObservable`
- `outputXX` is consistent with the `output()` API.
   - More clear semantics of a "declaration of an output" vs. a runtime conversion. `outputFromObservable` is a compiler construct.
   - enabling auto-completion as well / discoverability.
- `rxjsOutput` or `rxOutput` were options we considered— but it's ambiguous because it seemingly declared as different type of input. An RxJS output— but in practice the observable is _just_ the source.

We can also always iterate more on the naming here.

jelbourn reviewed (approved):
LGTM

Reviewed-for: public-api

devversion commented:
Passing TGP, except for two targets that need manual fixes in G3. Sent CLs to teams:

- cl/612832506
- cl/612832931

First one is unrelated to 3P, and second one is a subject to a faulty test asserting deep internal structure of `EventEmitter`—not something we should bother in terms of public API

devversion commented:
[Passing TGP](https://fusion2.corp.google.com/presubmit/611127431/OCL:611127431:BASE:613120011:1709716094375:6ad185d5;groups=Passing,PossiblyNewlyFailing/targets). 1 target failing, but seems quite unrelated. 

Caretaker note: Existing failures, TAP project of target seems red anyway.

pkozlowski-opensource commented:
This PR was merged into the repository by commit d4154f9e3c1719052485ce6d2f6c13d05c05e0d6.

TomieAi commented:
.required when? I love some vscode yelling I forgot something.

Flo0806 commented:
> .required when? I love some vscode yelling I forgot something.

The new `output` is the replacement for `EventEmitter()`. And it will be no good idea to have a `required` option. Think: You don't need the `click` event on every button. But if you **must** use it, it would just annoy the hell out of you. 

In the end: a `input` is different - here it makes sense to use required if needed, but not on a `output/EventEmitter`.

Greetings, Flo

blogcraft commented:
Why output() does not use signals like the new input()?

JoostK commented:
> Why output() does not use signals like the new input()?

It represents an _event_, i.e. an action at some point in time, rather than _state_, i.e. stored data. Signals correspond with state storage, not events.
